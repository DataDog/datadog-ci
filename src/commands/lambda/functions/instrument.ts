import {CloudWatchLogsClient} from '@aws-sdk/client-cloudwatch-logs'
import {
  LambdaClient,
  FunctionConfiguration as LFunctionConfiguration,
  UpdateFunctionConfigurationCommandInput,
} from '@aws-sdk/client-lambda'

import {
  API_KEY_ENV_VAR,
  CI_API_KEY_ENV_VAR,
  CI_SITE_ENV_VAR,
  ENVIRONMENT_ENV_VAR,
  SERVICE_ENV_VAR,
  SITE_ENV_VAR,
  VERSION_ENV_VAR,
} from '../../../constants'
import {isValidDatadogSite} from '../../../helpers/validation'

import {
  API_KEY_SECRET_ARN_ENV_VAR,
  ARM64_ARCHITECTURE,
  AWS_LAMBDA_EXEC_WRAPPER,
  AWS_LAMBDA_EXEC_WRAPPER_VAR,
  CAPTURE_LAMBDA_PAYLOAD_ENV_VAR,
  CI_API_KEY_SECRET_ARN_ENV_VAR,
  CI_KMS_API_KEY_ENV_VAR,
  CORECLR_ENABLE_PROFILING,
  CORECLR_PROFILER,
  CORECLR_PROFILER_PATH,
  DD_DOTNET_TRACER_HOME,
  DD_LAMBDA_EXTENSION_LAYER_NAME,
  DOTNET_TRACER_HOME_ENV_VAR,
  ENABLE_PROFILING_ENV_VAR,
  EXTENSION_LAYER_KEY,
  EXTRA_TAGS_ENV_VAR,
  FLUSH_TO_LOG_ENV_VAR,
  KMS_API_KEY_ENV_VAR,
  LAMBDA_HANDLER_ENV_VAR,
  LayerKey,
  LAYER_LOOKUP,
  LOG_LEVEL_ENV_VAR,
  MERGE_XRAY_TRACES_ENV_VAR,
  NODE_HANDLER_LOCATION,
  PROFILER_ENV_VAR,
  PROFILER_PATH_ENV_VAR,
  PYTHON_HANDLER_LOCATION,
  Runtime,
  RuntimeType,
  RUNTIME_LOOKUP,
  TRACE_ENABLED_ENV_VAR,
  APM_FLUSH_DEADLINE_MILLISECONDS_ENV_VAR,
  APPSEC_ENABLED_ENV_VAR,
} from '../constants'
import {FunctionConfiguration, InstrumentationSettings, LogGroupConfiguration, TagConfiguration} from '../interfaces'
import {calculateLogGroupUpdateRequest} from '../loggroup'
import {calculateTagUpdateRequest} from '../tags'

import {
  addLayerArn,
  findLatestLayerVersion,
  getLambdaFunctionConfigs,
  getLambdaFunctionConfigsFromRegex,
  getLayerArn,
  getLayers,
  isLayerRuntime,
  isSupportedRuntime,
} from './commons'
import {isExtensionCompatibleWithUniversalInstrumentation, isTracerCompatibleWithExtension} from './versionChecker'

export const getInstrumentedFunctionConfigs = async (
  lambdaClient: LambdaClient,
  cloudWatchLogsClient: CloudWatchLogsClient,
  region: string,
  functionARNs: string[],
  settings: InstrumentationSettings
): Promise<FunctionConfiguration[]> => {
  const lambdaFunctionConfigs = await getLambdaFunctionConfigs(lambdaClient, functionARNs)

  const configs: FunctionConfiguration[] = []
  for (const config of lambdaFunctionConfigs) {
    const functionConfig = await getInstrumentedFunctionConfig(
      lambdaClient,
      cloudWatchLogsClient,
      config,
      region,
      settings
    )

    configs.push(functionConfig)
  }

  return configs
}

export const getInstrumentedFunctionConfig = async (
  lambdaClient: LambdaClient,
  cloudWatchLogsClient: CloudWatchLogsClient,
  config: LFunctionConfiguration,
  region: string,
  settings: InstrumentationSettings
): Promise<FunctionConfiguration> => {
  const functionARN = config.FunctionArn!
  const runtime = config.Runtime
  if (!isSupportedRuntime(runtime)) {
    throw Error(`Can't instrument ${functionARN}, runtime ${runtime} not supported`)
  }

  const updateRequest = await calculateUpdateRequest(config, settings, region, runtime)
  let logGroupConfiguration: LogGroupConfiguration | undefined
  if (settings.forwarderARN !== undefined) {
    const logGroupName = `/aws/lambda/${config.FunctionName}`
    logGroupConfiguration = await calculateLogGroupUpdateRequest(
      cloudWatchLogsClient,
      logGroupName,
      settings.forwarderARN
    )
  }

  const tagConfiguration: TagConfiguration | undefined = await calculateTagUpdateRequest(lambdaClient, functionARN)

  return {
    functionARN,
    lambdaConfig: config,
    logGroupConfiguration,
    tagConfiguration,
    updateFunctionConfigurationCommandInput: updateRequest,
  }
}

export const getInstrumentedFunctionConfigsFromRegEx = async (
  lambdaClient: LambdaClient,
  cloudWatchLogsClient: CloudWatchLogsClient,
  region: string,
  pattern: string,
  settings: InstrumentationSettings
): Promise<FunctionConfiguration[]> => {
  const matchedFunctions = await getLambdaFunctionConfigsFromRegex(lambdaClient, pattern)
  const functionsToUpdate: FunctionConfiguration[] = []

  for (const config of matchedFunctions) {
    const functionConfig = await getInstrumentedFunctionConfig(
      lambdaClient,
      cloudWatchLogsClient,
      config,
      region,
      settings
    )
    functionsToUpdate.push(functionConfig)
  }

  return functionsToUpdate
}

export const calculateUpdateRequest = async (
  config: LFunctionConfiguration,
  settings: InstrumentationSettings,
  region: string,
  runtime: Runtime
) => {
  const oldEnvVars: Record<string, string> = {...config.Environment?.Variables}
  const changedEnvVars: Record<string, string> = {}
  const functionARN = config.FunctionArn

  const apiKey: string | undefined = process.env[CI_API_KEY_ENV_VAR] ?? process.env[API_KEY_ENV_VAR]
  const apiKeySecretArn: string | undefined = process.env[CI_API_KEY_SECRET_ARN_ENV_VAR]
  const apiKmsKey: string | undefined = process.env[CI_KMS_API_KEY_ENV_VAR]
  const site: string | undefined = process.env[CI_SITE_ENV_VAR]

  if (functionARN === undefined) {
    return undefined
  }

  const updateRequest: UpdateFunctionConfigurationCommandInput = {
    FunctionName: functionARN,
  }
  let needsUpdate = false
  const runtimeType = RUNTIME_LOOKUP[runtime]

  if (runtimeType === RuntimeType.CUSTOM) {
    if (settings.layerVersion !== undefined) {
      throw new Error(
        `Only the --extension-version argument should be set for the ${runtime} runtime. Please remove the --layer-version argument from the instrument command.`
      )
    }
  }

  // Update Python Handler
  if (runtimeType === RuntimeType.PYTHON && (settings.layerVersion !== undefined || settings.interactive)) {
    const expectedHandler = PYTHON_HANDLER_LOCATION
    if (config.Handler !== expectedHandler) {
      needsUpdate = true
      updateRequest.Handler = PYTHON_HANDLER_LOCATION
    }
  }

  // Update Node Handler
  if (runtimeType === RuntimeType.NODE && (settings.layerVersion !== undefined || settings.interactive)) {
    const expectedHandler = NODE_HANDLER_LOCATION
    if (config.Handler !== expectedHandler) {
      needsUpdate = true
      updateRequest.Handler = NODE_HANDLER_LOCATION
    }
  }

  // Update Env Vars
  if (runtimeType === RuntimeType.PYTHON || runtimeType === RuntimeType.NODE) {
    if (oldEnvVars[LAMBDA_HANDLER_ENV_VAR] === undefined) {
      needsUpdate = true
      changedEnvVars[LAMBDA_HANDLER_ENV_VAR] = config.Handler ?? ''
    }
  }

  // KMS > Secrets Manager > API Key
  if (apiKmsKey !== undefined && oldEnvVars[KMS_API_KEY_ENV_VAR] !== apiKmsKey) {
    needsUpdate = true
    changedEnvVars[KMS_API_KEY_ENV_VAR] = apiKmsKey
  } else if (apiKeySecretArn !== undefined && oldEnvVars[API_KEY_SECRET_ARN_ENV_VAR] !== apiKeySecretArn) {
    const isNode = runtimeType === RuntimeType.NODE
    const isSendingSynchronousMetrics = settings.extensionVersion === undefined && !settings.flushMetricsToLogs
    if (isSendingSynchronousMetrics && isNode) {
      throw new Error(
        '`apiKeySecretArn` is not supported for Node runtimes when using Synchronous Metrics. Use either `apiKey` or `apiKmsKey`.'
      )
    }
    needsUpdate = true
    changedEnvVars[API_KEY_SECRET_ARN_ENV_VAR] = apiKeySecretArn
  } else if (apiKey !== undefined && oldEnvVars[API_KEY_ENV_VAR] !== apiKey) {
    needsUpdate = true
    changedEnvVars[API_KEY_ENV_VAR] = apiKey
  }

  if (site !== undefined && oldEnvVars[SITE_ENV_VAR] !== site) {
    if (isValidDatadogSite(site)) {
      needsUpdate = true
      changedEnvVars[SITE_ENV_VAR] = site
    } else {
      throw new Error(
        'Warning: Invalid site URL. Must be either datadoghq.com, datadoghq.eu, us3.datadoghq.com, us5.datadoghq.com, ap1.datadoghq.com, or ddog-gov.com.'
      )
    }
  }
  if (site === undefined && oldEnvVars[SITE_ENV_VAR] === undefined) {
    needsUpdate = true
    changedEnvVars[SITE_ENV_VAR] = 'datadoghq.com'
  }

  const environmentVarsTupleArray: [keyof InstrumentationSettings, string][] = [
    ['apmFlushDeadline', APM_FLUSH_DEADLINE_MILLISECONDS_ENV_VAR],
    ['appsecEnabled', APPSEC_ENABLED_ENV_VAR],
    ['captureLambdaPayload', CAPTURE_LAMBDA_PAYLOAD_ENV_VAR],
    ['environment', ENVIRONMENT_ENV_VAR],
    ['extraTags', EXTRA_TAGS_ENV_VAR],
    ['mergeXrayTraces', MERGE_XRAY_TRACES_ENV_VAR],
    ['service', SERVICE_ENV_VAR],
    ['tracingEnabled', TRACE_ENABLED_ENV_VAR],
    ['version', VERSION_ENV_VAR],
  ]

  for (const [key, environmentVar] of environmentVarsTupleArray) {
    if (settings[key] !== undefined && oldEnvVars[environmentVar] !== settings[key]?.toString()) {
      needsUpdate = true
      changedEnvVars[environmentVar] = settings[key]!.toString()
    }
  }

  // Skip adding DD_FLUSH_TO_LOGS when using Extension
  const isUsingExtension = settings.extensionVersion !== undefined
  if (
    !isUsingExtension &&
    settings.flushMetricsToLogs !== undefined &&
    oldEnvVars[FLUSH_TO_LOG_ENV_VAR] !== settings.flushMetricsToLogs?.toString()
  ) {
    needsUpdate = true
    changedEnvVars[FLUSH_TO_LOG_ENV_VAR] = settings.flushMetricsToLogs.toString()
  }

  const newEnvVars = {...oldEnvVars, ...changedEnvVars}

  if (newEnvVars[LOG_LEVEL_ENV_VAR] !== settings.logLevel) {
    needsUpdate = true
    if (settings.logLevel) {
      newEnvVars[LOG_LEVEL_ENV_VAR] = settings.logLevel
    } else {
      delete newEnvVars[LOG_LEVEL_ENV_VAR]
    }
  }

  // Enable ASM
  if (settings['appsecEnabled'] === true) {
    newEnvVars[AWS_LAMBDA_EXEC_WRAPPER_VAR] = AWS_LAMBDA_EXEC_WRAPPER
  }

  let layerARNs = getLayers(config)
  const originalLayerARNs = layerARNs
  let needsLayerUpdate = false
  let layerOrTraceVersion: number | undefined
  if (isLayerRuntime(runtime)) {
    const lambdaLibraryLayerArn = getLayerArn(config, config.Runtime as LayerKey, region, settings)
    const lambdaLibraryLayerName = LAYER_LOOKUP[runtime as LayerKey]
    let fullLambdaLibraryLayerARN: string | undefined
    if (settings.layerVersion !== undefined || settings.interactive) {
      layerOrTraceVersion = settings.layerVersion
      if (settings.interactive && !settings.layerVersion) {
        layerOrTraceVersion = await findLatestLayerVersion(config.Runtime as LayerKey, region)
      }
      fullLambdaLibraryLayerARN = `${lambdaLibraryLayerArn}:${layerOrTraceVersion}`
    }
    layerARNs = addLayerArn(fullLambdaLibraryLayerARN, lambdaLibraryLayerName, layerARNs)
  }

  const lambdaExtensionLayerArn = getLayerArn(config, EXTENSION_LAYER_KEY as LayerKey, region, settings)
  let fullExtensionLayerARN: string | undefined
  let extensionVersion: number | undefined
  if (settings.extensionVersion !== undefined || settings.interactive) {
    extensionVersion = settings.extensionVersion
    if (settings.interactive && !settings.extensionVersion) {
      extensionVersion = await findLatestLayerVersion(EXTENSION_LAYER_KEY as LayerKey, region)
    }
    fullExtensionLayerARN = `${lambdaExtensionLayerArn}:${extensionVersion}`
  }
  layerARNs = addLayerArn(fullExtensionLayerARN, DD_LAMBDA_EXTENSION_LAYER_NAME, layerARNs)

  // Special handling for .NET and Java to support universal instrumentation
  if (runtimeType === RuntimeType.DOTNET || runtimeType === RuntimeType.JAVA) {
    if (layerOrTraceVersion && isExtensionCompatibleWithUniversalInstrumentation(runtimeType, extensionVersion)) {
      // If the user configures the trace version and the extension support univeral instrumenation
      // Then check whether the trace and extension are compatible with each other
      if (isTracerCompatibleWithExtension(runtimeType, layerOrTraceVersion)) {
        needsUpdate = true
        newEnvVars[AWS_LAMBDA_EXEC_WRAPPER_VAR] = AWS_LAMBDA_EXEC_WRAPPER
      } else {
        throw new Error(
          `For the ${runtime} runtime, the dd-trace version ${layerOrTraceVersion} is not compatible with the dd-extension version ${extensionVersion}`
        )
      }
    } else if (runtimeType === RuntimeType.DOTNET) {
      // If it is an old extension version or the trace version is null, leave it is as the old workflow
      if (
        !isExtensionCompatibleWithUniversalInstrumentation(runtimeType, extensionVersion) &&
        config.Architectures?.includes(ARM64_ARCHITECTURE)
      ) {
        throw new Error(
          'Instrumenting arm64 architecture is not supported for the given dd-extension version. Please choose the latest dd-extension version or use x86_64 architecture.'
        )
      } else {
        needsUpdate = true
        newEnvVars[ENABLE_PROFILING_ENV_VAR] = CORECLR_ENABLE_PROFILING
        newEnvVars[PROFILER_ENV_VAR] = CORECLR_PROFILER
        newEnvVars[PROFILER_PATH_ENV_VAR] = CORECLR_PROFILER_PATH
        newEnvVars[DOTNET_TRACER_HOME_ENV_VAR] = DD_DOTNET_TRACER_HOME
      }
    }
  }

  updateRequest.Environment = {
    Variables: newEnvVars,
  }

  if (originalLayerARNs.sort().join(',') !== layerARNs.sort().join(',')) {
    needsLayerUpdate = true
  }
  if (needsLayerUpdate) {
    needsUpdate = true
    updateRequest.Layers = layerARNs
  }

  layerARNs.forEach((layerARN) => {
    if (
      layerARN.includes(DD_LAMBDA_EXTENSION_LAYER_NAME) &&
      newEnvVars[API_KEY_ENV_VAR] === undefined &&
      newEnvVars[API_KEY_SECRET_ARN_ENV_VAR] === undefined &&
      newEnvVars[KMS_API_KEY_ENV_VAR] === undefined
    ) {
      throw new Error(
        `When 'extensionLayer' is set, ${CI_API_KEY_ENV_VAR}, ${CI_KMS_API_KEY_ENV_VAR}, or ${CI_API_KEY_SECRET_ARN_ENV_VAR} must also be set`
      )
    }
  })

  return needsUpdate ? updateRequest : undefined
}
