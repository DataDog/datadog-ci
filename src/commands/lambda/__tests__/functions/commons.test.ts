/* tslint:disable:no-string-literal */
import {
  DD_LAMBDA_EXTENSION_LAYER_NAME,
  EXTENSION_LAYER_KEY,
  EXTRA_TAGS_REG_EXP,
  GOVCLOUD_LAYER_AWS_ACCOUNT,
  LAMBDA_HANDLER_ENV_VAR,
  MERGE_XRAY_TRACES_ENV_VAR,
  Runtime,
  RUNTIME_LAYER_LOOKUP,
  TRACE_ENABLED_ENV_VAR,
} from '../../constants'
import {
  addLayerArn,
  collectFunctionsByRegion,
  getLayerArn,
  getRegion,
  sentenceMatchesRegEx,
  updateLambdaFunctionConfigs,
} from '../../functions/commons'
import {InstrumentCommand} from '../../instrument'
import {createCommand, makeMockCloudWatchLogs, makeMockLambda, mockAwsAccount} from '../fixtures'

describe('commons', () => {
  describe('addLayerArn', () => {
    test('adds layers and removes previous versions', () => {
      const runtime: Runtime = 'python3.9'
      const config = {
        Runtime: runtime,
      }
      let layerARNs = [
        'arn:aws:lambda:sa-east-1:464622532012:layer:Datadog-Python39:48',
        'arn:aws:lambda:sa-east-1:464622532012:layer:Datadog-Extension:10',
      ]
      const region = 'sa-east-1'
      const lambdaLibraryLayerName = RUNTIME_LAYER_LOOKUP[runtime]
      const fullLambdaLibraryLayerArn = getLayerArn(config, config.Runtime, region) + ':49'
      const fullExtensionLayerArn = getLayerArn(config, EXTENSION_LAYER_KEY as Runtime, region) + ':11'
      layerARNs = addLayerArn(fullLambdaLibraryLayerArn, lambdaLibraryLayerName, layerARNs)
      layerARNs = addLayerArn(fullExtensionLayerArn, DD_LAMBDA_EXTENSION_LAYER_NAME, layerARNs)

      expect(layerARNs).toEqual([
        'arn:aws:lambda:sa-east-1:464622532012:layer:Datadog-Python39:49',
        'arn:aws:lambda:sa-east-1:464622532012:layer:Datadog-Extension:11',
      ])
    })

    test('swaps layers if architecture is arm64', () => {
      const runtime: Runtime = 'python3.9'
      const config = {
        Architectures: ['arm64'],
        Runtime: runtime,
      }
      let layerARNs = [
        'arn:aws:lambda:sa-east-1:464622532012:layer:Datadog-Python39:49',
        'arn:aws:lambda:sa-east-1:464622532012:layer:Datadog-Extension:11',
      ]
      const region = 'sa-east-1'
      const lambdaLibraryLayerName = RUNTIME_LAYER_LOOKUP[runtime]
      const fullLambdaLibraryLayerArn = getLayerArn(config, config.Runtime, region) + ':49'
      const fullExtensionLayerArn = getLayerArn(config, EXTENSION_LAYER_KEY as Runtime, region) + ':11'
      layerARNs = addLayerArn(fullLambdaLibraryLayerArn, lambdaLibraryLayerName, layerARNs)
      layerARNs = addLayerArn(fullExtensionLayerArn, DD_LAMBDA_EXTENSION_LAYER_NAME, layerARNs)

      expect(layerARNs).toEqual([
        'arn:aws:lambda:sa-east-1:464622532012:layer:Datadog-Python39-ARM:49',
        'arn:aws:lambda:sa-east-1:464622532012:layer:Datadog-Extension-ARM:11',
      ])
    })
  })
  describe('collectFunctionsByRegion', () => {
    test('groups functions with region read from arn', () => {
      process.env = {}
      const command = createCommand(InstrumentCommand)
      const region = 'us-east-1'
      command['functions'] = [
        'arn:aws:lambda:us-east-1:123456789012:function:lambda-hello-world',
        'arn:aws:lambda:us-east-1:123456789012:function:another',
        'arn:aws:lambda:us-east-2:123456789012:function:third-func',
      ]

      expect(collectFunctionsByRegion(command['functions'], region)).toEqual({
        'us-east-1': [
          'arn:aws:lambda:us-east-1:123456789012:function:lambda-hello-world',
          'arn:aws:lambda:us-east-1:123456789012:function:another',
        ],
        'us-east-2': ['arn:aws:lambda:us-east-2:123456789012:function:third-func'],
      })
    })

    test('groups functions in the config object', () => {
      process.env = {}
      const command = createCommand(InstrumentCommand)
      const region = 'us-east-1'
      command['config']['functions'] = [
        'arn:aws:lambda:us-east-1:123456789012:function:lambda-hello-world',
        'arn:aws:lambda:us-east-1:123456789012:function:another',
        'arn:aws:lambda:us-east-2:123456789012:function:third-func',
      ]

      expect(collectFunctionsByRegion(command['config']['functions'], region)).toEqual({
        'us-east-1': [
          'arn:aws:lambda:us-east-1:123456789012:function:lambda-hello-world',
          'arn:aws:lambda:us-east-1:123456789012:function:another',
        ],
        'us-east-2': ['arn:aws:lambda:us-east-2:123456789012:function:third-func'],
      })
    })

    test('uses default region for functions not in arn format', () => {
      process.env = {}
      const command = createCommand(InstrumentCommand)
      command['functions'] = [
        'arn:aws:lambda:us-east-1:123456789012:function:lambda-hello-world',
        'arn:aws:lambda:*:123456789012:function:func-with-wildcard',
        'func-without-region',
        'arn:aws:lambda:us-east-2:123456789012:function:third-func',
      ]
      command['region'] = 'ap-south-1'

      expect(collectFunctionsByRegion(command['functions'], command['region'])).toEqual({
        'ap-south-1': ['arn:aws:lambda:*:123456789012:function:func-with-wildcard', 'func-without-region'],
        'us-east-1': ['arn:aws:lambda:us-east-1:123456789012:function:lambda-hello-world'],
        'us-east-2': ['arn:aws:lambda:us-east-2:123456789012:function:third-func'],
      })
    })

    test('fails to collect when there are regionless functions and no default region is set', () => {
      process.env = {}
      const command = createCommand(InstrumentCommand)
      command['functions'] = [
        'arn:aws:lambda:us-east-1:123456789012:function:lambda-hello-world',
        'arn:aws:lambda:*:123456789012:function:func-with-wildcard',
        'func-without-region',
        'arn:aws:lambda:us-east-2:123456789012:function:third-func',
      ]
      command['region'] = undefined
      command['config']['region'] = undefined
      let functionsGroup
      try {
        functionsGroup = collectFunctionsByRegion(command['functions'], command['region'])
      } catch (err) {
        // Do nothing
      }
      expect(functionsGroup).toBeUndefined()
    })
  })
  describe('getLayerArn', () => {
    const OLD_ENV = process.env
    beforeEach(() => {
      jest.resetModules()
      process.env = {}
    })
    afterAll(() => {
      process.env = OLD_ENV
    })

    test('gets sa-east-1 Lambda Extension layer ARN', async () => {
      const settings = {
        flushMetricsToLogs: false,
        layerAWSAccount: mockAwsAccount,
        mergeXrayTraces: false,
        tracingEnabled: false,
      }
      const region = 'sa-east-1'
      const layerArn = getLayerArn({}, EXTENSION_LAYER_KEY as Runtime, region, settings)
      expect(layerArn).toEqual(`arn:aws:lambda:${region}:${mockAwsAccount}:layer:Datadog-Extension`)
    })

    test('gets sa-east-1 arm64 Lambda Extension layer ARN', async () => {
      const config = {
        Architectures: ['arm64'],
      }
      const settings = {
        flushMetricsToLogs: false,
        layerAWSAccount: mockAwsAccount,
        mergeXrayTraces: false,
        tracingEnabled: false,
      }
      const region = 'sa-east-1'
      const layerArn = getLayerArn(config, EXTENSION_LAYER_KEY as Runtime, region, settings)
      expect(layerArn).toEqual(`arn:aws:lambda:${region}:${mockAwsAccount}:layer:Datadog-Extension-ARM`)
    })

    test('gets us-gov-1 gov cloud Lambda Extension layer ARN', async () => {
      const settings = {
        flushMetricsToLogs: false,
        layerAWSAccount: mockAwsAccount,
        mergeXrayTraces: false,
        tracingEnabled: false,
      }
      const region = 'us-gov-1'
      const layerArn = getLayerArn({}, EXTENSION_LAYER_KEY as Runtime, region, settings)
      expect(layerArn).toEqual(`arn:aws-us-gov:lambda:${region}:${GOVCLOUD_LAYER_AWS_ACCOUNT}:layer:Datadog-Extension`)
    })

    test('gets us-gov-1 gov cloud arm64 Lambda Extension layer ARN', async () => {
      const config = {
        Architectures: ['arm64'],
      }
      const settings = {
        flushMetricsToLogs: false,
        layerAWSAccount: mockAwsAccount,
        mergeXrayTraces: false,
        tracingEnabled: false,
      }
      const region = 'us-gov-1'
      const layerArn = getLayerArn(config, EXTENSION_LAYER_KEY as Runtime, region, settings)
      expect(layerArn).toEqual(
        `arn:aws-us-gov:lambda:${region}:${GOVCLOUD_LAYER_AWS_ACCOUNT}:layer:Datadog-Extension-ARM`
      )
    })

    test('gets sa-east-1 Node12 Lambda Library layer ARN', async () => {
      const runtime = 'nodejs12.x'
      const config = {
        Runtime: runtime,
      }
      const settings = {
        flushMetricsToLogs: false,
        layerAWSAccount: mockAwsAccount,
        mergeXrayTraces: false,
        tracingEnabled: false,
      }
      const region = 'sa-east-1'
      const layerArn = getLayerArn(config, config.Runtime as Runtime, region, settings)
      expect(layerArn).toEqual(`arn:aws:lambda:${region}:${mockAwsAccount}:layer:Datadog-Node12-x`)
    })

    test('gets sa-east-1 Python3.9 arm64 Lambda Library layer ARN', async () => {
      const runtime = 'python3.9'
      const config = {
        Architectures: ['arm64'],
        Runtime: runtime,
      }
      const settings = {
        flushMetricsToLogs: false,
        layerAWSAccount: mockAwsAccount,
        mergeXrayTraces: false,
        tracingEnabled: false,
      }
      const region = 'sa-east-1'
      const layerArn = getLayerArn(config, config.Runtime as Runtime, region, settings)
      expect(layerArn).toEqual(`arn:aws:lambda:${region}:${mockAwsAccount}:layer:Datadog-Python39-ARM`)
    })
    test('gets us-gov-1 Python37 gov cloud Lambda Library layer ARN', async () => {
      const runtime = 'python3.7'
      const config = {
        Runtime: runtime,
      }
      const settings = {
        flushMetricsToLogs: false,
        layerAWSAccount: mockAwsAccount,
        mergeXrayTraces: false,
        tracingEnabled: false,
      }
      const region = 'us-gov-1'
      const layerArn = getLayerArn(config, config.Runtime as Runtime, region, settings)
      expect(layerArn).toEqual(`arn:aws-us-gov:lambda:${region}:${GOVCLOUD_LAYER_AWS_ACCOUNT}:layer:Datadog-Python37`)
    })
    test('gets us-gov-1 Python39 gov cloud arm64 Lambda Library layer ARN', async () => {
      const runtime = 'python3.9'
      const config = {
        Architectures: ['arm64'],
        Runtime: runtime,
      }
      const settings = {
        flushMetricsToLogs: false,
        layerAWSAccount: mockAwsAccount,
        mergeXrayTraces: false,
        tracingEnabled: false,
      }
      const region = 'us-gov-1'
      const layerArn = getLayerArn(config, config.Runtime as Runtime, region, settings)
      expect(layerArn).toEqual(
        `arn:aws-us-gov:lambda:${region}:${GOVCLOUD_LAYER_AWS_ACCOUNT}:layer:Datadog-Python39-ARM`
      )
    })
  })
  describe('getRegion', () => {
    test('should return the expected region', () => {
      const functionARN = 'arn:aws:lambda:us-east-1:123456789012:function:lambda-hello-world'
      const expectedRegion = 'us-east-1'

      const region = getRegion(functionARN)
      expect(region).toBe(expectedRegion)
    })

    test('should return undefined if Function ARN does not contain the region', () => {
      const functionName = 'lambda-hello-world'

      const region = getRegion(functionName)
      expect(region).toBe(undefined)
    })
  })
  describe('sentenceMatchesRegEx', () => {
    const tags: [string, boolean][] = [
      ['not-complying:regex-should-fail', false],
      ['1first-char-is-number:should-fail', false],
      ['_also-not-complying:should-fail', false],
      ['complying_tag:accepted/with/slashes.and.dots,but-empty-tag', false],
      ['also_complying:success,1but_is_illegal:should-fail', false],
      ['this:complies,also_this_one:yes,numb3r_in_name:should-succeed,dots:al.lo.wed', true],
      ['complying_ip_address_4:192.342.3134.231', true],
      ['complying:alone', true],
      ['one_divided_by_two:1/2,one_divided_by_four:0.25,three_minus_one_half:3-1/2', true],
      ['this_is_a_valid_t4g:yes/it.is-42', true],
    ]
    test.each(tags)('check if the tags match the expected result from the regex', (tag, expectedResult) => {
      const result = !!sentenceMatchesRegEx(tag, EXTRA_TAGS_REG_EXP)
      expect(result).toEqual(expectedResult)
    })
  })

  describe('updateLambdaConfigs', () => {
    const OLD_ENV = process.env
    beforeEach(() => {
      jest.resetModules()
      process.env = {}
    })
    afterAll(() => {
      process.env = OLD_ENV
    })

    test('updates every lambda', async () => {
      const lambda = makeMockLambda({})
      const configs = [
        {
          functionARN: 'arn:aws:lambda:us-east-1:000000000000:function:autoinstrument',
          lambdaConfig: {
            FunctionArn: 'arn:aws:lambda:us-east-1:000000000000:function:autoinstrument',
            Handler: 'index.handler',
            Runtime: 'nodejs12.x',
          },
          lambdaLibraryLayerArn: 'arn:aws:lambda:us-east-1:464622532012:layer:Datadog-Node12-x',
          updateRequest: {
            Environment: {
              Variables: {
                [LAMBDA_HANDLER_ENV_VAR]: 'index.handler',
                [MERGE_XRAY_TRACES_ENV_VAR]: 'false',
                [TRACE_ENABLED_ENV_VAR]: 'false',
              },
            },
            FunctionName: 'arn:aws:lambda:us-east-1:000000000000:function:autoinstrument',
            Handler: '/opt/nodejs/node_modules/datadog-lambda-js/handler.handler',
            Layers: ['arn:aws:lambda:us-east-1:464622532012:layer:Datadog-Node12-x:22'],
          },
        },
      ]
      const cloudWatch = makeMockCloudWatchLogs({})

      await updateLambdaFunctionConfigs(lambda as any, cloudWatch as any, configs)
      expect(lambda.updateFunctionConfiguration).toHaveBeenCalledWith({
        Environment: {
          Variables: {
            [LAMBDA_HANDLER_ENV_VAR]: 'index.handler',
            [MERGE_XRAY_TRACES_ENV_VAR]: 'false',
            [TRACE_ENABLED_ENV_VAR]: 'false',
          },
        },
        FunctionName: 'arn:aws:lambda:us-east-1:000000000000:function:autoinstrument',
        Handler: '/opt/nodejs/node_modules/datadog-lambda-js/handler.handler',
        Layers: ['arn:aws:lambda:us-east-1:464622532012:layer:Datadog-Node12-x:22'],
      })
    })
  })
})
