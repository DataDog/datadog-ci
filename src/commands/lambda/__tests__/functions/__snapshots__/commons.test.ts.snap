// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`commons maskConfig should mask a Lambda config correctly 1`] = `
{
  "Architectures": [
    "x86_64",
  ],
  "CodeSize": 2275,
  "Environment": {
    "Variables": {
      "DD_API_KEY": "02**********33bd",
      "DD_LOG_LEVEL": "debug",
      "DD_SITE": "datadoghq.com",
    },
  },
  "FunctionArn": "arn:aws:lambda:us-east-1:123456789012:function:some-function",
  "FunctionName": "some-function",
  "Handler": "/path/handler.handler",
  "Layers": [
    {
      "Arn": "arn:aws:lambda:us-east-1:464622532012:layer:Datadog-Extension:43",
      "CodeSize": 13145076,
    },
    {
      "Arn": "arn:aws:lambda:us-east-1:464622532012:layer:Datadog-Node18-x:91",
      "CodeSize": 3614995,
    },
  ],
  "MemorySize": 1024,
  "Runtime": "nodejs18.x",
  "Timeout": 6,
}
`;

exports[`commons maskConfig should not modify config if env vars are missing 1`] = `
{
  "Architectures": [
    "x86_64",
  ],
  "CodeSize": 2275,
  "Environment": {},
  "FunctionArn": "arn:aws:lambda:us-east-1:123456789012:function:some-function",
  "FunctionName": "some-function",
  "Handler": "/path/handler.handler",
  "Layers": [
    {
      "Arn": "arn:aws:lambda:us-east-1:464622532012:layer:Datadog-Extension:43",
      "CodeSize": 13145076,
    },
    {
      "Arn": "arn:aws:lambda:us-east-1:464622532012:layer:Datadog-Node18-x:91",
      "CodeSize": 3614995,
    },
  ],
  "MemorySize": 1024,
  "Runtime": "nodejs18.x",
  "Timeout": 6,
}
`;
